src/org/traccar/notificators/NotificatorFirebase.java

```diff
@@ -46,29 +49,38 @@
         private String[] tokens;
         @JsonProperty("notification")
         private Notification notification;
+        @JsonProperty("data")
+        private Map<String, String> data;
     }
 
     public NotificatorFirebase() {
         key = Context.getConfig().getString("notificator.firebase.key");
+        isFormatDataOnly = Context.getConfig().getBoolean("notificator.firebase.dataOnly");
     }
 
     @Override
     public void sendSync(long userId, Event event, Position position) {
         final User user = Context.getPermissionsManager().getUser(userId);
         if (user.getAttributes().containsKey("notificationTokens")) {
 
-            Notification notification = new Notification();
-            notification.body = NotificationFormatter.formatShortMessage(userId, event, position).trim();
-
             Message message = new Message();
             message.tokens = user.getString("notificationTokens").split("[, ]");
-            message.notification = notification;
-
-            Context.getClient().target(URL).request()
+            message.data = NotificationFormatter.buildData(userId, event, position);
+            String body = NotificationFormatter.formatShortMessage(userId, event, position).trim();
+            if (isFormatDataOnly) {
+                message.data.put("body", body);
+            } else {
+                Notification notification = new Notification();
+                notification.body = body;
+                message.notification = notification;
+            }
+
+            ClientBuilder.newClient().target(URL).request()
                     .header("Authorization", "key=" + key)
                     .async().post(Entity.json(message), new InvocationCallback<Object>() {
                 @Override
                 public void completed(Object o) {
+                    LOGGER.info("Firebase notification sent to user: {}", userId);
```

Comment:
If you think we really need this, it should not be "info" level. Make it debug or lower.